### PR TITLE
etcd only nodes

### PR DESCRIPTION
**Pull request by galal-hussein**
## Design Takeaways

- Introduce a role flag to the server  process , and it can be set to either etcd or controlplane or both (default: controlplane,etcd)
- when role etcd is set to the role flag, the server process will:
  - generate the certs and bootstrap the cluster like it normally does
  - start the dynamic listener with all routes including the /db/info
  - start the agent process (connect kubelet/kubeproxy to the api)
  - start etcd (as a pod in case of rke2 and internally in case of k3s)
  - start an internal loadbalancer exactly like the agent with a prespecified port (6666)
- if  server role only is passed to the server process it will:
  - generate the certs and bootstrap the cluster like it normally does
  - start the dynamic listener with all routes including the /db/info
  - start the agent process (connect kubelet/kubeproxy to the api)
  - start all server components except etcd
- leverage taints for etcd nodes to work with SUC

###  ETCD nodes kubelet connecting to the API server

Server nodes will start either normally without args or with --role controlplane, in order for the server to join an etcd only node it should start with --server <etcd-ip>:<supervisor port>, once the server starts it will do the following:

 - parse the cli srever conf
 - create the data dir
 - grab the bootstrap data from the etcd node
 - sets the /etcd/serverip key and sets it to its node ip
 - start the server daemon and agent

63be342e-b990-48c8-b228-adc0e57fb922

### server nodes connecting to all etcd nodes in the cluster

So normally kubeapi connects to 127.0.0.1:2379 to connect to etcd on localhost, but since we are separating roles it needs access to all etcd nodes in the cluster, and to accomplish that I added special loadbalancer (etcdProxy) that first initialized with --server flag and when it connects to the first etcd node it will automatically and constantly query etcd memberlist and update the proxy endpoint without the need to restart kubeapi

### CODE REVIEW

File: pkg/cluster/cluster.go
Comment (brandond) [inline pkg/cluster/cluster.go:43]: What's the benefit of setting up a tunnel for this? There's some complicated logic in the etcd client libraries that it feels like we're bypassing by hiding the whole cluster behind a single proxy endpoint.

What does it give us over just passing Kubernetes the actual etcd endpoint client addresses and just letting core etcd client handle retries and faillover? Is it just the ability to add/remove cluster members at runtime without having to restart the apiserver?

File: pkg/cli/cmds/agent.go
Comment (brandond) [inline pkg/cli/cmds/agent.go:159]: What is this needed for?

File: pkg/cli/cmds/server.go
Comment (brandond) [inline pkg/cli/cmds/server.go:265]: Should this be a StringSliceFlag like the --disable list?

File: pkg/cli/cmds/server.go
Comment (briandowns) [inline pkg/cli/cmds/server.go:12]: Was this change intended?

File: pkg/cli/server/server.go
Comment (briandowns) [inline pkg/cli/server/server.go:308]: Is this the point that might benefit from being a Warn rather than an Error?

File: pkg/etcd/controller.go
Comment (briandowns) [inline pkg/etcd/controller.go:87]:
```suggestion
	address, ok := node.Annotations[NodeAddress] 
	if !ok {
	    return node, nil
	}
```

File: pkg/cluster/cluster.go
Comment (galal-hussein) [reply pkg/cluster/cluster.go:43]: yes exactly it mitigate the need for restarting the api everytime we add new etcd only node

File: pkg/cli/cmds/agent.go
Comment (galal-hussein) [reply pkg/cli/cmds/agent.go:159]: this is interesting workaround that I added, basically this sets the load balancer port instead of using a random port, I defaulted this value if not set to 6666 and this way it gets assigned to APIServerPort in server config and all kubeconfigs gets generated with this endpoint 127.0.0.1:6666 and also this gets assigned to the loadbalancer for api servers

File: pkg/cli/cmds/server.go
Comment (galal-hussein) [reply pkg/cli/cmds/server.go:265]: yes will fix, thanks

File: pkg/cli/server/server.go
Comment (galal-hussein) [reply pkg/cli/server/server.go:308]: no not really, will fix

File: pkg/cli/cmds/server.go
Comment (galal-hussein) [reply pkg/cli/cmds/server.go:12]: nope, that was a typo

File: pkg/cli/cmds/agent.go
Comment (brandond) [reply pkg/cli/cmds/agent.go:159]: As @erikwilson suggested in Slack, can we make this the same as the apiserver client port? It's only used on nodes without a local etcd server, so the port should be available.

File: pkg/agent/loadbalancer/loadbalancer_test.go
Comment (briandowns) [inline pkg/agent/loadbalancer/loadbalancer_test.go:108]: Can you define the `0` as a const with a descriptive name?

Comment (briandowns) [reply pkg/agent/loadbalancer/loadbalancer_test.go:108]: This would be dependent on: https://github.com/k3s-io/k3s/pull/2772/files#r555299396

File: pkg/cli/cmds/server.go
Comment (briandowns) [inline pkg/cli/cmds/server.go:267]: I wonder if it'd be a benefit to the user if the "valid items" section was listed in the format expected. ie "controlplane", "etcd", or "controlplane,etcd".

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:881]: Save a bunch of splitting and string converting, and just have the etcd key store a comma-separated list of server URLs since that's what we end up converting this to anyways.

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:905]: Just convert this to a https URI now since that's what we'll need later anyway.

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:890]:
```suggestion
			logrus.Warn("runtime is not ready yet")
```

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:885]: It appears that every control-plane server will, on startup, update this key to set themselves as the bootstrap endpoint for the control-plane load balancer. Is that intended?

The warnings in this goroutine should include some indication of where they come from, for example:
```
logrus.Warnf("Unable to register control-plane address with etcd: %v", err)
```

File: pkg/etcdctl/main.go
Comment (brandond) [inline pkg/etcdctl/main.go:51]: Kubernetes has only supported etcd api v3 for years so I think we should just hardcode v3. We will never need to support v2.

File: pkg/etcdctl/main.go
Comment (brandond) [inline pkg/etcdctl/main.go:67]: I think we have an existing function in the CLI wrapper for this. It might need to be modified so that you can use it outside the package, but use that instead of maintaining another copy here.

File: pkg/server/server.go
Comment (brandond) [inline pkg/server/server.go:449]: Use the const!
```suggestion
			if taint.Key == ETCDRoleLabelKey {
```

File: pkg/server/etcd.go
Comment (brandond) [inline pkg/server/etcd.go:84]: Do we need to remove the ControlPlane and Master role labels if this node is no longer running the control plane?

**Review by brandond — CHANGES_REQUESTED** [review]
The code has a lot of `if !serverConfig.DisableServer {} else {}` and places where disabling server seems to imply enabling etcd and so forth. I would like to see ControlConfig's DisableServer and DisableETCD fields switched to EnableServer and EnableEtcd, and the logic updated accordingly so that we're not testing a negative. This would improve code clarity a lot.

**Comment by brandond** [comment]
I don't see any code in here for handling removal of nodes from the etcd cluster when they no longer have the etcd role. Is that not supported at the moment?

I was halfway through a rewrite of the etc cluster controller to handle ensuring that the etcd cluster does not contain any members who aren't in the Kubernetes cluster with an etcd node-role label. I can continue to work on that once you merge this, if you want to leave that as out-of-scope for the moment.

**Comment by galal-hussein** [comment]
closed in favor of #2900

**Pull request closed without merging**
